File: generated_code/cpp/solution_3354_2.cpp
#include <string>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

class Solution {
public:
    string minimizeStringValue(string s) {
        // Frequency array to keep track of counts of each character already present in s
        vector<int> counts(26, 0);
        vector<int> question_mark_indices;
        
        // Count initial frequencies of fixed characters and locate '?' indices
        for (int i = 0; i < s.length(); ++i) {
            if (s[i] != '?') {
                counts[s[i] - 'a']++;
            } else {
                question_mark_indices.push_back(i);
            }
        }
        
        // If there are no '?' characters, return s as it is
        if (question_mark_indices.empty()) {
            return s;
        }
        
        // Min-heap to always pick the character with the minimum current frequency.
        // Pairs are stored as {frequency, char_index} to naturally handle 
        // the lexicographically smallest character tie-breaker.
        priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
        for (int i = 0; i < 26; ++i) {
            pq.push({counts[i], i});
        }
        
        // Greedily collect the best characters to fill the '?' slots.
        // This minimizes the overall cost contribution regardless of the replacement order.
        string replacements = "";
        for (int i = 0; i < question_mark_indices.size(); ++i) {
            auto [freq, ch_idx] = pq.top();
            pq.pop();
            
            replacements += (char)('a' + ch_idx);
            
            // Increment the frequency and push back into the heap
            pq.push({freq + 1, ch_idx});
        }
        
        // To guarantee the lexicographically smallest final string, 
        // the chosen characters must be placed in sorted order across the '?' indices.
        sort(replacements.begin(), replacements.end());
        
        // Replace the '?' tokens in s with the sorted replacement characters
        for (int i = 0; i < question_mark_indices.size(); ++i) {
            s[question_mark_indices[i]] = replacements[i];
        }
        
        return s;
    }
};

File: generated_code/python/solution_3354_2.py
from typing import List
import heapq

class Solution:
    def minimizeStringValue(self, s: str) -> str:
        # Count the current frequency of each character in the string
        freq = [0] * 26
        question_count = 0
        for ch in s:
            if ch == '?':
                question_count += 1
            else:
                freq[ord(ch) - ord('a')] += 1
        
        # If there are no '?' characters, return the original string
        if question_count == 0:
            return s
        
        # Min-heap to always pick the character with the minimum current frequency.
        # Elements are tuples: (frequency, character_index) to naturally prioritize
        # smaller frequencies first, and lexicographically smaller characters on ties.
        heap = [(freq[i], i) for i in range(26)]
        heapq.heapify(heap)
        
        # Collect the best characters to replace the '?' symbols
        chosen_chars = []
        for _ in range(question_count):
            count, idx = heapq.heappop(heap)
            chosen_chars.append(chr(ord('a') + idx))
            # Push the updated frequency back into the heap
            heapq.heappush(heap, (count + 1, idx))
            
        # To ensure the final string is lexicographically smallest overall,
        # the chosen characters must be placed in sorted order across the '?' positions.
        chosen_chars.sort()
        
        # Reconstruct the string by replacing '?' with the sorted chosen characters
        ans = []
        ptr = 0
        for ch in s:
            if ch == '?':
                ans.append(chosen_chars[ptr])
                ptr += 1
            else:
                ans.append(ch)
                
        return "".join(ans)